src/common/com/intellij/plugins/haxe/ide/annotator/HaxeSemanticAnnotator.java

```diff
@@ -188,624 +253,706 @@ private static boolean isFieldInitializedInTheConstructor(HaxeFieldModel field)
+    context) {
+    final List<HaxeParameterModel> parameters = Objects.requireNonNull(method.method).getParameters();
+
+    int parametersCount = parameters.size();
+    int argumentsCount = arguments.size();
+    int parameterIndex = 0;
+
+    for (int argumentIndex = 0; argumentIndex < argumentsCount; argumentIndex++) {
+      HaxeParameterModel parameter = parameterIndex >= parametersCount ? null : parameters.get(parameterIndex);
+      HaxeExpression argument = arguments.get(argumentIndex);
+
+      if (parameter == null) {
+        addExcessiveArgumentsError(arguments, argument);
+        return;
+      }
+
+      ResultHolder parameterType = method.getArguments().get(parameterIndex).getType();
+      ResultHolder argumentType = HaxeExpressionEvaluator.evaluate(argument, new HaxeExpressionEvaluatorContext(context, null)).result;
+
+      if (!parameterType.canAssign(argumentType)) {
+        int index = getIndexOfSuitableParameterAhead(parameters, parameterIndex, argumentType);
+        if (index != -1) {
+          parameterIndex = index;
+        } else {
+          annotateNotCompatibleTypes(argument, parameterType, argumentType, true);
+        }
+      }
+      parameterIndex++;
+    }
+
+    int requiredParametersCount = (int)parameters.stream()
+      .filter(parameter -> !(parameter.isOptional() || parameter.hasInitializer()))
+      .count();
+
+    if (requiredParametersCount > argumentsCount) {
+      String message = HaxeBundle.message("haxe.semantic.0.arguments.expected.got.1", requiredParametersCount, argumentsCount);
+      createErrorAnnotation(context, message);
+    }
   }
-}
 
-class InitVariableVisitor extends HaxeVisitor {
-  public boolean result = false;
+  private Annotation annotateNotCompatibleTypes(PsiElement annotationTarget,
+                                                ResultHolder destination,
+                                                ResultHolder source,
+                                                boolean addCastFix) {
+    String message = HaxeBundle
+      .message("haxe.semantic.incompatible.types", wrapWithHtmlError(source.toStringWithoutConstant()),
+               destination.toStringWithoutConstant());
+
 
-  private final String fieldName;
+    Annotation annotation = createErrorAnnotation(annotationTarget, message, true);
 
-  InitVariableVisitor(String fieldName) {
-    this.fieldName = fieldName;
+    if (addCastFix) {
+      annotation.registerFix(new HaxeAddCastFix(annotationTarget, source.getType(), destination.getType()));
+    }
+
+    return annotation;
   }
 
-  @Override
-  public void visitElement(PsiElement element) {
-    super.visitElement(element);
-    if (result) return;
-    if (element instanceof HaxeIdentifier || element instanceof HaxePsiToken || element instanceof HaxeStringLiteralExpression) return;
-    element.acceptChildren(this);
+  private String wrapWithHtmlError(String string) {
+    String color = UIUtil.isUnderDarcula() ? "FF6B68" : "red";
```

Comment:
This color should be in the settings dialog.  It's OK  to create a "technical debt" issue for this one, because it's going to take UI work that I'd rather see done in a separate PR.